File: src/com/maddyhome/idea/vim/helper/DigraphResult.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2021 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.helper;

import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class DigraphResult {
  public static final int RES_HANDLED = 0;
  public static final int RES_UNHANDLED = 1;
  public static final int RES_DONE = 3;
  public static final int RES_BAD = 4;

  static final DigraphResult HANDLED_DIGRAPH = new DigraphResult(RES_HANDLED, '?');
  static final DigraphResult HANDLED_LITERAL = new DigraphResult(RES_HANDLED, '^');
  static final DigraphResult UNHANDLED = new DigraphResult(RES_UNHANDLED);
  static final DigraphResult BAD = new DigraphResult(RES_BAD);

  private DigraphResult(int result) {
    this.result = result;
    stroke = null;
  }

  private DigraphResult(int result, char promptCharacter) {
    this.result = result;
    this.promptCharacter = promptCharacter;
    stroke = null;
  }

  private DigraphResult(@Nullable KeyStroke stroke) {
    result = RES_DONE;
    this.stroke = stroke;
  }

  public static DigraphResult done(@Nullable KeyStroke stroke) {
    return new DigraphResult(stroke);
  }

  public static DigraphResult handled(char promptCharacter) {
    return new DigraphResult(RES_HANDLED, promptCharacter);
  }

  public @Nullable KeyStroke getStroke() {
    return stroke;
  }

  public int getResult() {
    return result;
  }

  public char getPromptCharacter() {
    return promptCharacter;
  }

  private final int result;
  private final @Nullable KeyStroke stroke;
  private char promptCharacter;
}

File: vim-engine/src/main/kotlin/com/maddyhome/idea/vim/common/DigraphResult.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.common

sealed class DigraphResult(
  val codepoint: Int? = null,
  val promptCharacter: Char = 0.toChar(),
) {
  open class Handled(promptCharacter: Char) : DigraphResult(null, promptCharacter)
  data object HandledDigraph : Handled('?')
  data object HandledLiteral : Handled('^')
  class Done(codepoint: Int?) : DigraphResult(codepoint)
  data object Unhandled : DigraphResult()
  data object Bad : DigraphResult()

  companion object {
    fun done(codepoint: Int?): DigraphResult {
      // For some reason, vim does not let to insert char 10 as a digraph; it inserts 10 instead
      return Done(if (codepoint == 10) 0 else codepoint)
    }

    fun handled(promptCharacter: Char) = Handled(promptCharacter)
  }
}
